src/Pusher.php

```diff
@@ -745,4 +751,48 @@ public function notify($interests, $data = array(), $debug = false)
 
         return false;
     }
+
+    /**
+     * Verify that a webhook actually came from Pusher, and marshals them into a PusherWebhook object.
+     *
+     * @param array  $headers a array of headers from the request (for example, from getallheaders())
+     * @param string $body    the body of the request (for example, from file_get_contents('php://input'))
+     *
+     * @return PusherWebhook object with the properties time_ms (an int) and events (an array of event objects)
+     */
+    public function parse_webhook($headers, $body)
+    {
+        if (!$this->valid_signature($headers, $body)) {
+            return false;
+        }
+        $decoded_events = array();
+        $decoded_json = json_decode($body);
+        $webhookobj = new PusherWebhook();
+        $webhookobj->time_ms = $decoded_json->time_ms;
+        $webhookobj->events = $decoded_json->events;
+
+        return $webhookobj;
+    }
+
+    /**
+     * Verify that a given Pusher Signature is valid.
+     *
+     * @param array  $headers a array of headers from the request (for example, from getallheaders())
+     * @param string $body    the body of the request (for example, from file_get_contents('php://input'))
+     *
+     * @return bool true if signature is correct.
+     */
+    public function valid_signature($headers, $body)
+    {
+        $x_pusher_key = $headers['X-Pusher-Key'];
+        $x_pusher_signature = $headers['X-Pusher-Signature'];
+        if ($x_pusher_key == $this->settings['auth_key']) {
+            $expected = hash_hmac('sha256', $body, $this->settings['secret']);
+
+            return $expected === $x_pusher_signature;
+        }
+        $this->log('Invalid Webhook signature.', array(), LogLevel::WARNING);
```

Comment:
I think we should avoid logging errors from synchronous code in order to give the user of the library more control over what to log. I think we should return an exception from the highest level possible, in this case this method could return `false`, and the function above could throw an exception.

Reply:
I also think we could give more helpful error messages here, e.g. like https://github.com/pusher/pusher-http-ruby/blob/master/lib/pusher/webhook.rb#L105.

If the exception needs to be thrown from here to show a more useful error message, then I would rename it `ensure_valid_signature`, not return anything, and just throw an exception if invalid.

Reply:
yep!